internal/builtin/providers/tf/provider_functions.go

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+// Copyright (c) 2023 HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package tf
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/hcl/v2/hclwrite"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/opentofu/opentofu/internal/providers"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// "decode_tfvars"
+// "encode_tfvars"
+// "encode_expr"
+
+// decodeTFVarsFunc decodes a TFVars file content into a cty object
+type decodeTFVarsFunc struct{}
+
+func (f *decodeTFVarsFunc) Name() string {
+	return "decode_tfvars"
+}
+
+func (f *decodeTFVarsFunc) GetFunctionSpec() providers.FunctionSpec {
+	params := []providers.FunctionParameterSpec{
+		{
+			Name:              "content",
+			Type:              cty.String,
+			Description:       "TFVars file content to decode",
+			DescriptionFormat: providers.TextFormattingPlain,
+		},
+	}
+	return providers.FunctionSpec{
+		Parameters:        params,
+		Return:            cty.DynamicPseudoType,
+		Summary:           "Decode a TFVars file content into an object",
+		Description:       "provider::terraform::decode_tfvars decodes a TFVars file content into an object",
+		DescriptionFormat: providers.TextFormattingPlain,
+	}
+}
+
+var FailedToDecodeError = errors.New("failed to decode tfvars content")
+
+func wrapDiagErrors(m error, diag hcl.Diagnostics) error {
+	//Prepend the main error
+	errs := append([]error{m}, diag.Errs()...)
+	return errors.Join(errs...)
+}
+
+func (f *decodeTFVarsFunc) Call(args []cty.Value) (cty.Value, error) {
+	varsFileContent := args[0].AsString()
+	schema, diag := hclsyntax.ParseConfig([]byte(varsFileContent), "", hcl.Pos{Line: 0, Column: 0})
+	if schema == nil || diag.HasErrors() {
+		return cty.NullVal(cty.DynamicPseudoType), wrapDiagErrors(FailedToDecodeError, diag)
+	}
+	attrs, diag := schema.Body.JustAttributes()
+	// Check if there are any errors.
+	// attrs == nil does not mean that there are no attributes, attrs - is still initialized as an empty map
+	if attrs == nil || diag.HasErrors() {
+		return cty.NullVal(cty.DynamicPseudoType), wrapDiagErrors(FailedToDecodeError, diag)
+	}
+	vals := make(map[string]cty.Value)
+	for name, attr := range attrs {
+		val, diag := attr.Expr.Value(nil)
+		if diag.HasErrors() {
+			return cty.NullVal(cty.DynamicPseudoType), wrapDiagErrors(FailedToDecodeError, diag)
+		}
+		vals[name] = val
+	}
+	return cty.ObjectVal(vals), nil
+}
+
+// encodeTFVarsFunc encodes an object into a string with the same format as a TFVars file
+type encodeTFVarsFunc struct{}
+
+func (f *encodeTFVarsFunc) Name() string {
+	return "encode_tfvars"
+}
+
+func (f *encodeTFVarsFunc) GetFunctionSpec() providers.FunctionSpec {
+	params := []providers.FunctionParameterSpec{
+		{
+			Name: "input",
+			// The input type is determined at runtime
+			Type:              cty.DynamicPseudoType,
```

Comment:
This is a nitpick, but I was wondering if we can put a more specific type here in the input type, as we know that the input should be an `object`

Reply:
I went through the docs for cty types. There doesn't seem to be any other way to have an object type without knowing it's concrete type at compilation time.
If you can find more descriptive type for our usecase, that would definitely be better.

Reply:
I looked at this a little bit but didn't found anything, so let's drop this